legacy/logclient/persistance.go

```diff
@@ -0,0 +1,130 @@
+/*
+Copyright 2021 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package logclient
+
+import (
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/sirupsen/logrus"
+)
+
+// PersistentLoggingInfo holds peristand stats about the behavior of the auditor.
+// Since the auditor is a continuously running program, this data will begin
+// collection upon startup and lost on shutdown. Such data is only collected when
+// invoked with the --verbose flag.
+type PersistentLoggingInfo struct {
+	verboseLogging bool             // Determines if the user passed the --verbose flag, allowing persistent logs.
+	mutex          sync.Mutex       // Required to grab the queries after each measurement window.
+	measurement    QueryMeasurement // The measurement structure holding the amount of times GCR was queried.
+}
+
+// QueryMeasurement holds the logging information about number of queries per measurment window.
+type QueryMeasurement struct {
+	startTime  time.Time
+	recorded   bool
+	numQueries uint64
+}
+
+// Debug only logs the given message if the verboseLogging option is true.
+func Debug(args ...interface{}) {
+	if PersistentAuditInfo.verboseLogging {
+		logrus.Debug(args...)
+	}
+}
+
+// CountGCRQuery adds the given number to the persistent audit logs only
+// if verboseLogging is enabled.
+func CountGCRQuery(n int) {
+	// Only capture queries when verboseLogging is enabled
+	if !PersistentAuditInfo.verboseLogging {
+		return
+	}
+	// Grab and the lock to prevent race conditions.
+	PersistentAuditInfo.mutex.Lock()
+	defer PersistentAuditInfo.mutex.Unlock()
+	// Determin if the current measurement has already been recorded.
+	if !PersistentAuditInfo.measurement.recorded {
+		PersistentAuditInfo.measurement.numQueries += uint64(n)
+		return
+	}
+	// Create a new measurement.
+	PersistentAuditInfo.measurement = QueryMeasurement{
+		startTime:  time.Now(),
+		recorded:   false,
```

Comment:
I guess `recorded` here means whether we used this `PersistentAuditInfo.measurement` information to increment `numQueries`?

Reply:
recorded is now removed, and is not needed to manage state correctly. Resolving comment.